### DISCUSSION TITLE
Node-locked license issues with Docker

### DISCUSSION THREAD

**Discussion by diracpulse** [Xilinx/finn#571] [Q&A] [answered]
Hello everyone. I have a ZCU102 and I am trying to run the end2end examples. I am able to synthesize the code for the PYNQ devices, but when i try to do the same for the ZCU-102 I get the following error : Synthesis failed, no bitfile found. Check logs under /tmp/finn_dev_name/vivado_zynq_proj__gr37nr1

I have looked in the logfiles (and single-module sub-logfiles) produced and it looks like the feature "Synthesis" for the ZCU102 MPSoC is not available. The error there is the following :

Starting synth_design Attempting to get a license for feature 'Synthesis' and/or device 'xczu9eg' WARNING: [Common 17-348] Failed to get the license for feature 'Synthesis' and/or device 'xczu9eg'. Explanation: The license feature Synthesis could not be found. Resolution: Check the status of your licenses in the Vivado License Manager. For debug help search Xilinx Support for "Licensing FAQ". 24 Infos, 1 Warnings, 0 Critical Warnings and 1 Errors encountered. synth_design failed ERROR: [Common 17-345] A valid license was not found for feature 'Synthesis' and/or device 'xczu9eg'. Please run the Vivado License Manager for assistance in determining which features and devices are licensed for your system. Resolution: Check the status of your licenses in the Vivado License Manager. For debug help search Xilinx Support for "Licensing FAQ". If you are using a license server, verify that the license server is up and running a version of the xilinx daemon that is compatible with the version of Xilinx software that you are using. Please note that Vivado 2017.3 and later requires upgrading your license server tools to the Flex 11.14.1 tools. Please confirm with your license admin that the correct version of the license server tools are installed. INFO: [Common 17-206] Exiting Vivado at Thu Mar 10 09:29:23 2022...

All the sub-logfiles state the same error.

As far as I can tell from using the combination of ZCU102 and Vivado/Vitis HLS I am able to synthesize and deploy to the device.
I have checked which licenses are installed on my machine, there is no feature called 'Synthesis' but all the other licenses seem to be ok.

Sorry for the beginner question here and thanks in advance for the answer.

**Comment by fpjentzsch** [comment, marked as answer]
Hi, the ZCU102 is not covered by the free WebPack license. Since FINN runs Vivado from within the Docker container, you'll need to pass your license information (e.g. a file or path to a license server) to this container. See https://finn.readthedocs.io/en/latest/getting_started.html#vivado-vitis-license

**Reply by diracpulse** [reply]
Hello, I have tried it. I have used the command : 
`export FINN_DOCKER_EXTRA=" -v /home/<user>/.Xilinx/:/home/<user>/.Xilinx/ -e XILINXD_LICENSE_FILE=/home/<user>/.Xilinx/Xilinx.lic "` in the same console that I use to run FINN. Unluckily I am getting the same error. 

In the `/home/<user>/.Xilinx` folder there surely is a license file.

**Reply by btwbtw01** [reply]
Hi, @vittorioserra @fpjentzsch 
I'm not sure that it is the license problem. I didn't set the path, but I can run 3-build-accelerator-with-finn example and run on ZCU104 board successfully. However, I run VGG16 model, it shows the error. 
![VGG16_error_compile](https://user-images.githubusercontent.com/85877169/158005736-fc0b880e-5bb8-43c3-9304-0d2079ec3bf2.PNG)
![VGG16_error_compile_detail](https://user-images.githubusercontent.com/85877169/158005742-3a786ad1-9994-4716-9073-7e3721bfc576.PNG)
Although "ERROR: [Common 17-345]" is license problem, I can run example, but can't run the model that I trained. I'm not sure it is without license.

**Reply by diracpulse** [reply]
Hello @btwbtw01 @fpjentzsch, 

what do your secondary logfiles say?

**Reply by diracpulse** [reply]
Hello @fpjentzsch,

I have double checked that the environment variable is set correctly inside the docker container, and have checked that I can reach the folder containing the Vivado license files from it. 

An `echo $XILINXD_LICENSE_FILE` returns `/home/<user>/.Xilinx/Xilinx.lic`, so this seems to be set fine. Furthermore, I can access `/home/<user>/.Xilinx/` from the docker container, and I am able to see the license file in it. In the license file there is an increment for the XCZU9EG and the ES1 and ES2 extensions are present.

The license I am using is a **node-locked** one.

There is a really interesting turn anyway. By running Vivado in tcl-model inside the Docker container, and feeding it the following command `report_environment -file xinfo.txt` I was able to generate an info-file regarding the environment, including the licenses. 

Interestingly enough, when the file is generated inside the Docker container, I can see that there are no licenses for the XCZU9EG devices.  The following two pictures are screenshots of this file, just to make things clearer.

![image](https://user-images.githubusercontent.com/5065137/159698422-8aadea89-c5c6-4602-b39a-32f23a7b150f.png)
![image](https://user-images.githubusercontent.com/5065137/159698473-d34fdf8e-34c9-4d32-9554-bc2c5569cf74.png)

When running the same command outside of the docker container, the "Host Id Matches" column contains "Yes" values for the XCZU9EG devices. It is worthy  of notice that the Host-Id is the same both in and outside of the Docker container.

**Reply by fpjentzsch** [reply]
That is weird. I never tried a node-locked license myself, but found this thread: https://github.com/Xilinx/finn/pull/280#issuecomment-830986100
To what kind of host ID is your license locked? A FLEXID dongle?

**Reply by diracpulse** [reply]
It is locked to the MAC address of my PC's eth0 interface. We were able to find a solution to the problem.

It is necessary to give the same Host-ID to the docker container too, hence the MAC-Address of the docker container needs to be set. To do it, modify the docker run call in `run-docker.sh` by adding the same MAC-address as the Host-ID's to the command.

The line will look something like this : `DOCKER_EXEC="docker run -t --rm $DOCKER_INTERACTIVE --tty --init --mac-address <Host-ID-MAC-Address> "`

**Reply by diracpulse** [reply]
Another solution, which would be more advantageous in order to connect to the board in later steps, would be to add `--network=host`in the `DOCKER_EXEC`command

**Reply by fpjentzsch** [reply]
Makes sense, thanks for posting your solution! I'll mark this discussion as answered and we might add these instructions to our "Getting Started" guide.

**Reply by jay8078** [reply]
I tried adding --network=host and it's not working. Any suggestions, please?

**Reply by pkdeep** [reply]
One can also try to copy the node-locked license file inside docker ("/tmp/home_dir/.Xilinx"). That worked for me.
